jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -191,40 +191,64 @@ public int getMaxCapacity()
     public RetainableByteBuffer acquire(int size, boolean direct)
     {
         RetainedBucket bucket = bucketFor(size, direct);
+
+        // No bucket, return non-pooled.
         if (bucket == null)
-            return newRetainableByteBuffer(size, direct, this::removed);
+            return newRetainableByteBuffer(size, direct, null);
+
+        // Try to acquire a pooled entry.
         Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().acquire();
+        if (entry != null)
+        {
+            RetainableByteBuffer buffer = entry.getPooled();
+            ((Buffer)buffer).acquire();
+            return buffer;
+        }
 
-        RetainableByteBuffer buffer;
+        return newRetainableByteBuffer(bucket.getCapacity(), direct, buffer -> reserve(bucket, buffer, direct));
+    }
+
+    private void reserve(RetainedBucket bucket, RetainableByteBuffer buffer, boolean direct)
+    {
+        // Discard the buffer if maxMemory is exceeded.
+        if (isMaxMemoryExceeded(bucket, direct))
+            return;
```

Comment:
This is potentially the wrong buffer (size) to discard.  If there are other buckets of different sizes with buffers that have been idle longer, then they should be discarded rather than a buffer of a size that has just been actively used.

Reply:
I guess this hypothesis needs to be verified.
On a normal load where all buffers are somehow equally used, there is no "right" size.
You will be looking at other sizes and they may also be equally actively used, complicating the algorithm for little or no benefit.

Reply:
The whole design of this pool is based on the assumption that we don't know the size of buffer that will be requested. Hence we have multiple different sizes of bucket.  We allow different buffer sizes to be configured for input, output and headers. Also there are used APIs to request buffers of particular sizes.

I think it is there other way around, that any assumption that there mid of sizes requested will not change over time.

Reply:
@gregw how is a high concurrent server and its applications suddenly decide to not use anymore a certain size of buffers? 
How all threads and all applications agree to a "change in size frequency"?

I don't think ABBP was ever designed with "the ability to adapt to changes in size frequency", so that was never a goal and nor it is now.

The design based on buckets is to avoid having a bucket for each different capacity, and improve pooling and reuse, not "changes in size frequency".

Reply:
@sbordet it could be lots of reasons:
- An application may have different types of traffic at different times of the day. 
- A deployment might use different types of buffers whilst initializing/idle vs when up and running and receiving traffic.
- A server might be in some kind of fail over scenario where it normally does not handle some type of traffic until some failure event. 
- JMX can be used to adjust configuration of a running server.
- A DOS attacker might flood a server with requests of a specific type that use a specific buffer size

The current ABBP certainly has the ability to adapt to changes in size frequency.  It specifically has been design to look for the least recently used buffers to be evicted.
It is this PR that proposes removing that ability, so I think you need to show that the size frequency will never EVER change on any deployment.  If the size frequency does change, then we will be effectively without a buffer pool and we have seen how that leads to OOM failure.  So this PR has to prove that is not a risk.

Ultimately, if adapting was not a concern, then why evict at all?  Just fill up and be done with it!

Reply:
> Ultimately, if adapting was not a concern, then why evict at all?

Because we wanted to constrain the memory pooled to a configurable amount.
I think it was a bad decision when we first introduced that feature to run around the buckets to try to evict.

My previous comments about the CometD benchmark show clearly how the unconstrained pool consumes 34 MiB, but constraining it at 4 MiB still has a very good pool hit rate.
This is important in containerized environments.

I'm fine having an alternative, slower, implementation that "adapts to changes in size frequency" for those applications that depending on the time of the day will request different buffer sizes, or if you connect to the production systems and change via JMX the buffer sizes.
But I think for normally behaving applications, and for the common cases, and as the default, we should have a simple and fast pool, rather than one that tries to be too smart (an approach we tried already and failed).

I'll try to collect information from our websites and report back.

Reply:
> @sbordet The whole design of this pool is based on the assumption that we don't know the size of buffer that will be requested. Hence we have multiple different sizes of bucket. We allow different buffer sizes to be configured for input, output and headers. Also there are used APIs to request buffers of particular sizes.

Don't forget that Websocket has typically been the one that pushes these assumptions about ByteBufferPool to the breaking point.

Reply:
@sbordet I think both this pool and the one from the other PR are going to perform mostly the same.    Expanding the pool should happen rarely, so the cost of scanning the buckets for the most-idle-bucket is going to be rarely used and is a simple iteration anyway.  This pool does a bit more work on every release (checking the maxSize), whilst the pool in the other PR typically does less work on every release (just checking a single atomic boolean) and will commonly not have to check the size.

So I don't think you can assume the other PR is slower.  @lorban can you benchmark both these PRs at some various pool sizes.  I bet there is nothing between them.

Moreover, I'm sure the normal/usual/common case is for the buffer size not to vary over time.  But we have to deal with the uncommon case, especially if that can be forced from a remote client, as if we end up with a pool full of the wrong bucket size, then we will fail with OOM when presented load at a different bucket size.    

I don't think it is satisfactory to have a default pool that can potentially be vulnerable to a serious attack/accidental DOS and if/when it happens say "you should have configured the other pool type".   If this pool really is faster, then we should say: "hey if you want to run a bit faster and are sure your buffer sizes do not vary, then use this pool variant".

Reply:
This PR also checks a single atomic on every `release()`.

Look at my comment on the other PR.

Reply:
I'm resolving this conversation as it is so long, it is making it hard to review the code.  Plus we have moved some way since it started.